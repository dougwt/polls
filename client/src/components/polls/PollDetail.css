.container {
  margin-top: 4em;
  margin-bottom: 4em;
}

.card-title {
  /* font-style: italic; */
}

.choices {
  padding-top: 2em;
  padding-bottom: 2em;
}
