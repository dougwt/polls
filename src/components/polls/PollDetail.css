.container {
  margin-top: 4em;
  margin-bottom: 4em;
}

.card-title {
  /* font-style: italic; */
}

.choices {
  padding-top: 2em;
  padding-bottom: 2em;
}

.choices label,
.choices .col {
  width: 100%;
}

.choices label:hover {
  color: #333;
  font-size: 1.1em;
}
