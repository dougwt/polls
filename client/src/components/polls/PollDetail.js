import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Card, Input, Button, Row } from 'react-materialize';
import { votePoll } from '../../actions';
import { Pie } from 'react-chartjs-2';

import './PollDetail.css';

const defaultPoll = {
  choices: [0, 1, 2, 3, 4, 5],
  question: 'Who is your favorite Starfleet captain?',
  choice_1: 'Kirk (TOS)',
  choice_2: 'Picard (TNG)',
  choice_3: 'Sisko (DS9)',
  choice_4: 'Janeway (Voyager)',
  choice_5: 'Archer (Enterprise)',
  choice_6: 'Lorca (Discovery)'
};

const onVoteSubmit = (pollId, votePoll, disabled) => {
  const checked = document.querySelector('Input[name="choices"]:checked');
  const choiceId = checked ? checked.value : undefined;

  if (!disabled && pollId && choiceId) {
    votePoll(pollId, choiceId);
  }
};

export const PollDetail = ({
  auth,
  poll,
  formValues = defaultPoll,
  disabled = false,
  votePoll,
  waiting
}) => {
  return (
    <div className="container">
      <Card className="darken-1" title={formValues.question}>
        {poll && poll.respondents.includes(auth._id)
          ? renderResults(poll)
          : renderForm(poll, formValues, disabled, votePoll, waiting)}
      </Card>
    </div>
  );
};
PollDetail.propTypes = {
  auth: PropTypes.object,
  poll: PropTypes.object,
  formValues: PropTypes.object,
  disabled: PropTypes.bool,
  votePoll: PropTypes.func,
  waiting: PropTypes.bool
};

function renderForm(poll, formValues, disabled, votePoll, waiting) {
  return (
    <form
      className="poll"
      onSubmit={event => {
        event.preventDefault();
        onVoteSubmit(poll._id, votePoll, disabled);
      }}
    >
      {/* TODO: add spacing here via CSS */}

      <div className="choices">{renderChoices(poll, formValues)}</div>

      <Row className="center-align">
        {renderAsyncButton(waiting, disabled, event => {
          event.preventDefault();
          onVoteSubmit(poll._id, votePoll, disabled);
        })}
      </Row>
    </form>
  );
}

function renderResults(poll) {
  const data = { datasets: [{ data: [], backgroundColor: [] }], labels: [] };

  const colors = ['red', 'blue', 'green', 'yellow', 'orange', 'purple'];

  poll.choices.forEach((choice, i) => {
    data.datasets[0].data.push(choice.votes);
    data.datasets[0].backgroundColor.push(colors[i]);
    data.labels.push(choice.text);
  });

  const options = {
    legend: { position: 'bottom' }
  };

  return <Pie data={data} options={options} />;
}

function renderChoices(poll, values) {
  function renderChoice(key, value, label) {
    return (
      <Row key={key}>
        <Input name="choices" type="radio" value={value} label={label} />
      </Row>
    );
  }

  let choices = [];

  if (!poll) {
    // If poll is undefined it has not been saved, so use fake ids
    for (
      let choice = 1;
      choice <= (values.choices ? values.choices.length : 0);
      choice++
    ) {
      let field = `choice_${choice}`;
      if (values[field]) {
        choices.push(renderChoice(choice, choice.toString(), values[field]));
      }
    }
  } else {
    // If defined, we can use real ids from the existing poll
    poll.choices.forEach(choice => {
      choices.push(renderChoice(choice._id, choice._id, choice.text));
    });
  }

  return choices;
}

function renderAsyncButton({ waiting, disabled, onClick }) {
  if (waiting) {
    return (
      <Button className="teal" disabled={true}>
        Saving...
      </Button>
    );
  }

  return (
    <Button className="teal" disabled={disabled} onClick={onClick}>
      Vote
    </Button>
  );
}
renderAsyncButton.propTypes = {
  disabled: PropTypes.bool.isRequired,
  onClick: PropTypes.func.isRequired,
  waiting: PropTypes.bool.isRequired
};

function mapStateToProps(state) {
  return {
    auth: state.auth,
    error: state.poll.error,
    waiting: state.poll.waiting
  };
}

export default connect(mapStateToProps, { votePoll })(PollDetail);
