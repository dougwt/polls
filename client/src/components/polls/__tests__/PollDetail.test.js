import React from 'react';
import { shallow } from 'enzyme';
import PollDetail from '../PollDetail';

const wrapper = shallow(<PollDetail />);
describe('PollDetail', () => {
  it('renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  // xit('contains a PollDetailForm', () => {
  //   // console.log(wrapper.debug());
  //   expect(wrapper.find('PollDetailForm').exists()).toBe(true);
  // });
});
