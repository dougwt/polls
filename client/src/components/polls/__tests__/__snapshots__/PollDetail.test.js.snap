// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PollDetail renders properly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Unknown
    disabled={false}
    formValues={
        Object {
            "choices": Array [
              0,
              1,
              undefined,
              undefined,
              undefined,
              undefined,
            ],
            "question": "Who is your favorite Starfleet captain?",
          }
    }
    votePoll={[Function]}
    waiting={false}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <Card
        className="darken-1"
        title="Who is your favorite Starfleet captain?"
>
        <form
                className="poll"
                onSubmit={[Function]}
        >
                <div
                        className="choices"
                >
                        
                </div>
                <Row
                        className="center-align"
                >
                        <Button
                                className="teal"
                                disabled={undefined}
                                node="button"
                                onClick={undefined}
                        >
                                Vote
                        </Button>
                </Row>
        </form>
</Card>,
      "className": "container",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": <form
          className="poll"
          onSubmit={[Function]}
>
          <div
                    className="choices"
          >
                    
          </div>
          <Row
                    className="center-align"
          >
                    <Button
                              className="teal"
                              disabled={undefined}
                              node="button"
                              onClick={undefined}
                    >
                              Vote
                    </Button>
          </Row>
</form>,
        "className": "darken-1",
        "title": "Who is your favorite Starfleet captain?",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div
              className="choices"
>
              
</div>,
            <Row
              className="center-align"
>
              <Button
                            className="teal"
                            disabled={undefined}
                            node="button"
                            onClick={undefined}
              >
                            Vote
              </Button>
</Row>,
          ],
          "className": "poll",
          "onSubmit": [Function],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [],
              "className": "choices",
            },
            "ref": null,
            "rendered": Array [],
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": <Button
                className="teal"
                disabled={undefined}
                node="button"
                onClick={undefined}
>
                Vote
</Button>,
              "className": "center-align",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": "Vote",
                "className": "teal",
                "disabled": undefined,
                "node": "button",
                "onClick": undefined,
              },
              "ref": null,
              "rendered": "Vote",
              "type": [Function],
            },
            "type": [Function],
          },
        ],
        "type": "form",
      },
      "type": [Function],
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <Card
          className="darken-1"
          title="Who is your favorite Starfleet captain?"
>
          <form
                    className="poll"
                    onSubmit={[Function]}
          >
                    <div
                              className="choices"
                    >
                              
                    </div>
                    <Row
                              className="center-align"
                    >
                              <Button
                                        className="teal"
                                        disabled={undefined}
                                        node="button"
                                        onClick={undefined}
                              >
                                        Vote
                              </Button>
                    </Row>
          </form>
</Card>,
        "className": "container",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": <form
            className="poll"
            onSubmit={[Function]}
>
            <div
                        className="choices"
            >
                        
            </div>
            <Row
                        className="center-align"
            >
                        <Button
                                    className="teal"
                                    disabled={undefined}
                                    node="button"
                                    onClick={undefined}
                        >
                                    Vote
                        </Button>
            </Row>
</form>,
          "className": "darken-1",
          "title": "Who is your favorite Starfleet captain?",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <div
                className="choices"
>
                
</div>,
              <Row
                className="center-align"
>
                <Button
                                className="teal"
                                disabled={undefined}
                                node="button"
                                onClick={undefined}
                >
                                Vote
                </Button>
</Row>,
            ],
            "className": "poll",
            "onSubmit": [Function],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [],
                "className": "choices",
              },
              "ref": null,
              "rendered": Array [],
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "children": <Button
                  className="teal"
                  disabled={undefined}
                  node="button"
                  onClick={undefined}
>
                  Vote
</Button>,
                "className": "center-align",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "children": "Vote",
                  "className": "teal",
                  "disabled": undefined,
                  "node": "button",
                  "onClick": undefined,
                },
                "ref": null,
                "rendered": "Vote",
                "type": [Function],
              },
              "type": [Function],
            },
          ],
          "type": "form",
        },
        "type": [Function],
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "disableLifecycleMethods": true,
  },
}
`;
